Azure Speech Integration

KanaHayama opened the pull request:
Hi,
I found this plugin while I was trying to create one myself to help me read papers and books. Luckily, I came across this plugin, so I didn’t have to build it from scratch, and I really enjoy using it. I tried the Windows adapter that includes Azure Speech and found its quality to be much better than the default TTS engine. However, that adapter has an issue where it only speaks a small portion of the text before stopping.

Since I have some experience using Azure Speech in other projects, I decided to integrate it here, and this is the result of my implementation. Although most of the code was generated with the help of an LLM, I’ve done my best to keep the design clean and have thoroughly tested it. Below are some key points of my implementation—please help review and merge it:

+ I chose to use the Azure Speech WebSocket v2 API instead of the JavaScript SDK, because I initially tried to use the new Input Text Streaming feature, which isn’t available in the JavaScript SDK. In the end, I couldn’t successfully utilize that feature, but I still think avoiding a strict dependency on the SDK is beneficial for future compatibility with other TTS engines. So I kept this design.
+ The TTS engine implementation is mostly self-contained in a single TypeScript file, with only minimal modifications to the plugin itself where necessary for integration. These include changes to the TTS engine selection UI. However, since I tried to minimize changes to the plugin, it might not fully integrate with other parts, such as the favorite voice feature, which I haven’t tested.
+ I manually implemented input text streaming and audio playback streaming, allowing it to read long papers continuously without delay or early termination. The input text streaming is implemented to reduce the cost of TTS by not sending all the text at once, and to circumvent the TTS duration limit, which has a maximum of 10 minutes.
+ I chose the OGG audio format to reduce network bandwidth.
+ I added several settings to the preferences page, including key, region, language, and voice. To address potential key confidentiality concerns, I also added the ability to read the key and region from environment variables.
+ I have only tested it on my Windows machine, but I feel compatibility should not be an issue, as it doesn’t rely on any third-party code.
+ The pitch and speed cannot be adjusted yet. Azure Speech does not provide a global setting for this. The only way to adjust pitch and speed is through constructing SSMLs on the fly, which can be tricky, and based on my past experience, the resulting audio often sounds inconsistent. So it is questionable how to do it well.
+ Azure TTS is billed by input characters. Currently, 1 million characters cost about $15, which is relatively expensive. It might be a good idea to warn users to be mindful of their usage in the UI.
+ Currently, it is hard to let the plugin read a textbook with Speak From Here, as creating the argument for the speak method takes a long time. So I haven't yet fully reached my goal. This may require a breaking change to the interface.

If you need any further adjustments, please let me know. Thank you.

KanaHayama committed:
add source map

KanaHayama committed:
Add basic Azure Speech support

KanaHayama committed:
Streaming audio play

KanaHayama committed:
Add text streaming

ImperialSquid commented:
Hi @KanaHayama 

Thanks so much for all the effort you've put into this, it's a massive help!

Just fyi, it may take me some time to review everything since there's a lot here and other life stuff is fairly busy at the moment.

KanaHayama committed:
Add Azure Speech volume and rate control

KanaHayama committed:
Minor code style adjustments

+ Fix incorrectly adjusted indentation
+ Adjust Azure export order
+ Change single quotes to double quotes
+ Convert region value to lowercase

KanaHayama committed:
Minor adjustments

KanaHayama committed:
fix: repopulate Azure UI from cache on prefs reload

KanaHayama committed:
fix: prevent undefined shown in Azure volume/rate on first open

Wenlin88 commented:
Really looking forward to getting this feature up and running! I can also help if there is anything where I could assist with this feature for ZoTTS.

KanaHayama committed:
add configurable playback preload amount setting

KanaHayama commented:
Hi, for those who want to try out this update, here is a built version of the plugin in its current state.
[zo-tts with Azure.zip](https://github.com/user-attachments/files/24425299/zo-tts.with.Azure.zip)

v4u6h4n commented:
@KanaHayama 

Hey there, just wondering how difficult it would be to use your work to use a local llm? I just got [kokoro](https://github.com/remsky/Kokoro-FastAPI) container setup and it works really well.

KanaHayama commented:
@v4u6h4n My modification calls the Azure Speech Web API. That API may be quite different from a generic LLM-based speech-to-text API. Other parts, such as segmenting and buffering, might be reusable for other streaming backends.

v4u6h4n commented:
Thanks for getting back to me :-) Yeah I did a little bit of digging, and I think Kokoro uses the OpenAI API. I still have to get around to learning Javascript so itll have to wait for now, although I may be able to borrow some code from the ReadAloud browser extension if I'm lucky.

EDIT: Looks like the new built in Zotero TTS might be able to work with Kokoro!
